src/k8s/pkg/k8sd/api/certificates_refresh.go

```diff
@@ -0,0 +1,281 @@
+			key:          &certificates.KubeletClientKey,
+		},
+		{
+			name:        fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", req.Seed),
+			commonName:  "system:kube-proxy",
+			usages:      []certificatesv1.KeyUsage{certificatesv1.UsageDigitalSignature, certificatesv1.UsageKeyEncipherment, certificatesv1.UsageClientAuth},
+			signerName:  "k8sd.io/kube-proxy-client",
+			certificate: &certificates.KubeProxyClientCert,
+			key:         &certificates.KubeProxyClientKey,
+		},
+	} {
+		csr := csr
+		g.Go(func() error {
+			csrPEM, keyPEM, err := pkiutil.GenerateCSR(
+				pkix.Name{
+					CommonName:   csr.commonName,
+					Organization: csr.organization,
+				},
+				2048,
+				csr.hostnames,
+				csr.ips,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to generate CSR for %s: %w", csr.name, err)
+			}
+
+			if _, err = client.CertificatesV1().CertificateSigningRequests().Create(ctx, &certificatesv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: csr.name,
+				},
+				Spec: certificatesv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Usages:     csr.usages,
+					SignerName: csr.signerName,
+				},
+			}, metav1.CreateOptions{}); err != nil {
+				return fmt.Errorf("failed to create CSR for %s: %w", csr.name, err)
+			}
+
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+
+					watcher, err := client.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: csr.name}))
+					if err != nil {
+						log.WithValues("csr", csr.name).V(1).Error(err, "failed to watch CSR")
+						time.Sleep(3 * time.Second)
+						continue
+					}
+
+					defer watcher.Stop()
+
+					watchClosed := false
+					for !watchClosed {
+						select {
+						case <-ctx.Done():
+							return nil
+						case evt, ok := <-watcher.ResultChan():
+							if !ok {
+								log.WithValues("csr", csr.name).V(1).Info("watch closed")
+								watchClosed = true
+							}
+
+							request, ok := evt.Object.(*certificatesv1.CertificateSigningRequest)
+							if !ok {
+								log.WithValues("csr", csr.name).Error(fmt.Errorf("expected a CertificateSigningRequest but received %#v", evt.Object), "unexpected object")
+								continue
+							}
+
+							approved, err := isCertificateSigningRequestApproved(request)
+							if err != nil {
+								return fmt.Errorf("csr is in an invalid state: %w", err)
+							}
+
+							if approved && isCertificateSigningRequestIssued(request) {
+								if _, _, err = pkiutil.LoadCertificate(string(request.Status.Certificate), ""); err != nil {
+									log.WithValues("csr", csr.name).Error(err, "CertificateSigningRequest failed")
+									return fmt.Errorf("CertificateSigningRequest failed: %w", err)
+								}
+								*csr.certificate = string(request.Status.Certificate)
+								*csr.key = string(keyPEM)
+								return nil
+							}
+						}
+					}
+					time.Sleep(3 * time.Second)
+				}
+
+			}
```

Comment:
this is a bit messy code, which can lead to undesired bugs down the line. a few notes:

- avoid `time.Sleep(3 * time.Second)` as that blocks even if the context cancels in the meantime. _if you do that_, go for `select` with `<-time.After(3*.time.Second)` 
- this would be written better as a helper function that recreates the watcher if it fails. as written, you are moving code in very deep indent depths. that's not a problem by itself, but you can reduce it. for example, you could do|

```go
select {
  case <-ctx.Done():
    // exit
  default:
    // code goes here
}
```

as

```go
select {
  case <-ctx.Done():
    // exit
  default:
}

// code goes here
```